components/org.wso2.carbon.identity.application.authenticator.iwa/src/main/java/org/wso2/carbon/identity/application/authenticator/iwa/IWAAuthenticationUtil.java

```diff
@@ -0,0 +1,268 @@
+     * @throws LoginException
+     */
+    private static GSSCredential createServerCredentials(CallbackHandler callbackHandler)
+            throws PrivilegedActionException, LoginException {
+        // authenticate to the AD (Kerberos Server)
+        LoginContext loginContext = new LoginContext(IWAConstants.SERVER, callbackHandler);
+        loginContext.login();
+
+        if (log.isDebugEnabled()) {
+            log.debug("Pre-authentication successful for with Kerberos Server.");
+        }
+        // create server credentials from pre authentication with the AD
+        return createCredentialsForSubject(loginContext.getSubject());
+    }
+
+
+    /**
+     * Create GSSCredential as Subject
+     *
+     * @param subject login context subject
+     * @return GSSCredential
+     * @throws PrivilegedActionException
+     */
+    private static GSSCredential createCredentialsForSubject(final Subject subject) throws PrivilegedActionException {
+        final PrivilegedExceptionAction<GSSCredential> action =
+                new PrivilegedExceptionAction<GSSCredential>() {
+                    public GSSCredential run() throws GSSException {
+                        return gssManager.createCredential(null, GSSCredential.INDEFINITE_LIFETIME,
+                                IWAServiceDataHolder.getSpnegoOid(), GSSCredential.ACCEPT_ONLY);
+                    }
+                };
+
+        if (log.isDebugEnabled()) {
+            Set<Principal> principals = subject.getPrincipals();
+            String principalName = null;
+            if (principals != null) {
+                principalName = principals.toString();
+            }
+            log.debug("Creating gss credentials as principal : " + principalName);
+        }
+        return Subject.doAs(subject, action);
+    }
+
+
+    /**
+     * Create call back handler using given username and password
+     *
+     * @param username
+     * @param password
+     * @return CallbackHandler
+     */
+    private static CallbackHandler getUsernamePasswordHandler(final String username, final String password) {
+        final CallbackHandler handler = new CallbackHandler() {
+            public void handle(final Callback[] callback) {
+                for (int i = 0; i < callback.length; i++) {
+                    Callback currentCallBack = callback[i];
+                    if (currentCallBack instanceof NameCallback) {
+                        final NameCallback nameCallback = (NameCallback) currentCallBack;
+                        nameCallback.setName(username);
+                    } else if (currentCallBack instanceof PasswordCallback) {
+                        final PasswordCallback passCallback = (PasswordCallback) currentCallBack;
+                        passCallback.setPassword(password.toCharArray());
+                    } else {
+                        log.error("Unsupported Callback i = " + i + "; class = " + currentCallBack.getClass().getName());

```

Comment:
'i' doesnt seem to add any value to the error message. Fix with https://github.com/wso2-extensions/identity-local-auth-iwa/pull/36